### PR TITLE
Adds interface for configuration and sampling types

### PR DESCRIPTION
**Pull request by vulder**

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 21
           

Complexity increasing per file
==============================
- varats-core/varats/base/configuration.py  3
- tests/base/test_commandline_option.py  3
- tests/mapping/test_configuration_map.py  2
- tests/test_helper_config.py  4
- varats-core/varats/base/commandline_option.py  3
- varats-core/varats/mapping/configuration_map.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 21
           

Complexity increasing per file
==============================
- varats-core/varats/base/configuration.py  3
- tests/mapping/test_configuration_map.py  2
- varats-core/varats/base/commandline_option.py  3
- tests/base/test_commandline_option.py  3
- tests/test_helper_config.py  4
- varats-core/varats/mapping/configuration_map.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 25
           

Complexity increasing per file
==============================
- varats-core/varats/base/configuration.py  3
- varats-core/varats/base/commandline_option.py  3
- tests/mapping/test_configuration_map.py  2
- tests/test_helper_config.py  4
- varats-core/varats/mapping/configuration_map.py  4
- tests/base/test_commandline_option.py  3
         

Complexity decreasing per file
==============================
+ varats-core/varats/base/sampling_method.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 23
           

Complexity increasing per file
==============================
- tests/base/test_commandline_option.py  3
- tests/mapping/test_configuration_map.py  2
- tests/test_helper_config.py  4
- varats-core/varats/mapping/configuration_map.py  4
- varats-core/varats/base/commandline_option.py  3
- varats-core/varats/base/configuration.py  3
         

Complexity decreasing per file
==============================
+ varats-core/varats/base/sampling_method.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 33
           

Complexity increasing per file
==============================
- tests/mapping/test_configuration_map.py  2
- varats-core/varats/base/configuration.py  3
- tests/base/test_sampling_method.py  1
- varats-core/varats/base/commandline_option.py  3
- tests/base/test_commandline_option.py  3
- tests/test_helper_config.py  4
- varats-core/varats/mapping/configuration_map.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 33
           

Complexity increasing per file
==============================
- tests/base/test_sampling_method.py  1
- tests/base/test_commandline_option.py  3
- varats-core/varats/mapping/configuration_map.py  4
- varats-core/varats/base/configuration.py  3
- tests/test_helper_config.py  4
- tests/mapping/test_configuration_map.py  2
- varats-core/varats/base/commandline_option.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 17
           

Complexity increasing per file
==============================
- varats-core/varats/mapping/configuration_map.py  4
- varats-core/varats/base/commandline_option.py  3
- tests/test_helper_config.py  4
- varats-core/varats/base/configuration.py  3
- tests/base/test_commandline_option.py  3
- tests/mapping/test_configuration_map.py  2
- tests/base/test_sampling_method.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 17
           

Complexity increasing per file
==============================
- varats-core/varats/mapping/configuration_map.py  4
- tests/base/test_sampling_method.py  1
- tests/test_helper_config.py  4
- tests/base/test_commandline_option.py  3
- varats-core/varats/base/configuration.py  3
- varats/varats/tools/driver_casestudy.py  1
- tests/mapping/test_configuration_map.py  2
- varats-core/varats/base/commandline_option.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 23
           

Complexity increasing per file
==============================
- varats-core/varats/mapping/configuration_map.py  4
- varats/varats/tools/driver_casestudy.py  1
- tests/test_helper_config.py  4
- tests/base/test_configuration.py  1
- varats-core/varats/base/commandline_option.py  3
- tests/base/test_sampling_method.py  1
- tests/mapping/test_configuration_map.py  2
- tests/base/test_commandline_option.py  3
- varats-core/varats/base/configuration.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 19
           

Complexity increasing per file
==============================
- varats-core/varats/mapping/configuration_map.py  4
- tests/test_helper_config.py  4
- tests/mapping/test_configuration_map.py  2
- tests/base/test_sampling_method.py  1
- varats-core/varats/base/configuration.py  3
- tests/base/test_configuration.py  1
- tests/base/test_commandline_option.py  3
- varats-core/varats/base/commandline_option.py  3
- varats/varats/tools/driver_casestudy.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=6230787&bid=20304466)

File: varats-core/varats/base/configuration.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    @abc.abstractproperty
+    def name(self) -> str:
+        """The option name, refering to the feature from which this options
+        stems."""
+        raise NotImplementedError  # pragma: no cover
+
+    @abc.abstractproperty
+    def value(self) -> tp.Any:
+        """Current set value of the option."""
```
Comment (boehmseb) [inline varats-core/varats/base/configuration.py]:
```suggestion
        """Currently set value of the option."""
```

File: varats-core/varats/base/sampling_method.py [hunk 2]
```diff
@@ -1,18 +1,177 @@
...
-    uniform = 1
-    half_norm = 2
 
-    # TODO (sattlerf): refactor to correct base class
+class SamplingMethodBase(tp.Generic[SamplingMethodSubType], abc.ABC):
+    """
+    Represents the sampling configuration added as a base class to all sampling
+    methods.
 
+    This class is designed make the configuration of a sampling method
```
Comment (boehmseb) [inline varats-core/varats/base/sampling_method.py]:
```suggestion
    This class is designed to make the configuration of a sampling method
```

File: varats-core/varats/base/sampling_method.py [hunk 3]
```diff
@@ -1,18 +1,177 @@
...
+        Returns: SamplingConfiguration as a string
+        """
+        partial_config = self._extend_config()
+        partial_config[self.CONFIG_TYPE_NAME] = self.name()
+
+        return str(partial_config)
+
+    def _extend_config(self) -> tp.Dict[str, tp.Any]:  # pylint: disable=R0201
+        """Returns a configuration dict with config values from the sub class
+        that should be persisted."""
```
Comment (boehmseb) [inline varats-core/varats/base/sampling_method.py]:
```suggestion
        """
        Returns a configuration dict with config values from the sub class
        that should be persisted.
        
        Implementations in subclasses should always call 
        `super()._extend_config()` first.
        """
```

File: varats-core/varats/base/sampling_method.py [hunk 4]
```diff
@@ -1,18 +1,177 @@
...
+class NormalSamplingMethod(SamplingMethodBase['NormalSamplingMethod']):
+    """Abstract base class for normal sampling methods that sample following a
+    certain probability distribution."""
+
+    def __init__(self) -> None:
+        pass
+
+    def _extend_config(self) -> tp.Dict[str, tp.Any]:
+        """Returns a configuration dict with config values from the sub class
+        that should be persisted."""
```
Comment (boehmseb) [inline varats-core/varats/base/sampling_method.py]:
```suggestion
        """
        Returns a configuration dict with config values from the sub class
        that should be persisted.
        
        Implementations in subclasses should always call 
        `super()._extend_config()` first.
        """
```

File: varats-core/varats/base/sampling_method.py [hunk 5]
```diff
@@ -21,25 +180,166 @@ def gen_distribution_function(self) -> tp.Callable[[int], np.ndarray]:
...
+    """Abstract base class for feature-sampling methods that sample
+    configurations from a feature model based on different sampling
+    strategies."""
+
+    def __init__(self, solver: Solver = Solver.NoSolver) -> None:
+        self.__solver = solver
+
+    def _extend_config(self) -> tp.Dict[str, tp.Any]:
+        """Returns a configuration dict with config values from the sub class
+        that should be persisted."""
```
Comment (boehmseb) [inline varats-core/varats/base/sampling_method.py]:
```suggestion
        """
        Returns a configuration dict with config values from the sub class
        that should be persisted.
        
        Implementations in subclasses should always call 
        `super()._extend_config()` first.
        """
```

File: varats-core/varats/mapping/configuration_map.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+from varats.base.configuration import Configuration
+from varats.base.version_header import VersionHeader
+from varats.utils.exceptions import ConfigurationMapConfigIDMissmatch
+
+LOG = logging.getLogger(__name__)
+
+
+class ConfigurationMap():
+    """A configuration map builds a relation between an unique ID and the
```
Comment (boehmseb) [inline varats-core/varats/mapping/configuration_map.py]:
```suggestion
    """A configuration map builds a relation between a unique ID and the
```

**Review by boehmseb — CHANGES_REQUESTED** [review]

File: tests/base/test_configuration.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    def test_equality_different(self) -> None:
+        """Test to compare config option to to each other."""
+        config_option_int = ConfigurationOptionTestImpl("foo", 42)
+        config_option_bool = ConfigurationOptionTestImpl("foo", False)
+
+        self.assertFalse(config_option_int == config_option_bool)
+        self.assertTrue(config_option_int != config_option_bool)
+
+    def test_equality_othertypes(self) -> None:
+        """Test to compare config option to to each other."""
```
Comment (ChristianKaltenecker) [inline tests/base/test_configuration.py]:
```suggestion
        """Test to compare config option to each other."""
```

**Review by boehmseb — APPROVED** [review]

File: tests/test_helper_config.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        test_config.add_config_option(ConfigurationOptionTestImpl("bar", False))
+        test_config.add_config_option(
+            ConfigurationOptionTestImpl("bazz", "bazz-value")
+        )
+        return test_config
+
+    @staticmethod
+    def create_configuration_from_str(config_str: str) -> 'Configuration':
+        """
+        Creates a `Configuration` from it's string representation.
```
Comment (ChristianKaltenecker) [inline tests/test_helper_config.py]:
```suggestion
        Creates a `Configuration` from its string representation.
```

File: tests/test_helper_config.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+        This function is the inverse to `dump_to_string` to reparse a
+        configuration dumpred previously.
```
Comment (ChristianKaltenecker) [inline tests/test_helper_config.py]:
```suggestion
        configuration dumped previously.
```

File: varats-core/varats/base/configuration.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+class Configuration():
+    """Represents a specific configuration of a project, e.g., encapsulating
+    static and run-time options on how the project should be build."""
+
+    @staticmethod
+    @abc.abstractmethod
+    def create_configuration_from_str(config_str: str) -> 'Configuration':
+        """
+        Creates a `Configuration` from it's string representation.
```
Comment (ChristianKaltenecker) [inline varats-core/varats/base/configuration.py]:
```suggestion
        Creates a `Configuration` from its string representation.
```

**Review by ChristianKaltenecker — APPROVED** [review]

**Pull request merged**